rdkit/Chem/Pharm3D/EmbedLib.py

```diff
@@ -1259,6 +1364,309 @@ def ComputeChiralVolume(mol, centerIdx, confId=-1):
+    self.maxEmbeddings = maxEmbeddings
+    self.maxConformers = maxConformers
+    self.minFeaturesMatched = minFeaturesMatched
+
+  def __str__(self):
+    c = "ph4-constrained" if self.usePh4Constraints else "unconstrained"
+    e = "ETKDG" if self.useETKDG else "DG"
+    m = "minimized" if self.minimize else "unminimized"
+    mappings = "maxMappings%i" % self.maxMappings
+    embeddings = "maxEmbeddings%i" % self.maxEmbeddings
+    conformers = "maxConformers%i" % self.maxConformers
+    features = "minFeatures%i" % self.minFeaturesMatched
+    return "_".join([c, e, m, mappings, embeddings, conformers, features])
+
+
+class Ph4ConstrainedEmbedder:
+  def __init__(self, ligand, pharmacophore, settings):
+    self.ligand = ligand
+    self.pharmacophore = pharmacophore
+    self.settings = settings
+
+  def _match_pharmacophore(self, featFactory):
+    if self.settings.partialMatching:
+      canMatch, allMatches, pcophore = MatchPharmacophoreToMol(
+          self.ligand, featFactory, self.pharmacophore,
+          partialMatching=self.settings.partialMatching)
+      self.pharmacophore = pcophore
+    else:
+      canMatch, allMatches = MatchPharmacophoreToMol(
+          self.ligand, featFactory, self.pharmacophore,
+          partialMatching=False)
+
+    if not canMatch or (len(self.pharmacophore.getFeatures()) <
+        self.settings.minFeaturesMatched):
+      raise EmbedException(ExitState.CANNOT_MATCH_MOL)
+    return allMatches
+
+  def _recursive_map(self, matches, mat, pharmacophore):
+    """
+    Returns all valid mappings (after distance constraints are applied) while
+    allowing for partial matching of pharmacophore features.
+    """
+    if len(pharmacophore.getFeatures()) < self.settings.minFeaturesMatched:
+      return  # not enough features matched
+
+    explicit = False
+    if isinstance(pharmacophore, Pharmacophore.ExplicitPharmacophore):
+      explicit = True
+
+    for (_, _, mapping, _) in MatchPharmacophores(matches, mat, pharmacophore,
+        use2DLimits=True, mol=self.ligand, useDownsampling=False):
+      matches = [list(x.GetAtomIds()) for x in mapping]
+      self.mappings.append((matches, pharmacophore))
+      if len(self.mappings) > self.settings.maxMappings:
+        return
+
+    if not self.settings.partialMatching:
+      return  # stop after the first attempt
+
+    if len(self.mappings) == 0:
+      feats = pharmacophore.getFeatures()
+      for i in range(len(feats)):
+        featcopy = list(feats)
+        matchcopy = list(matches)
+        featcopy.pop(i)
+        matchcopy.pop(i)
+
+        if explicit:
+          radiicopy = list(pharmacophore.getRadii())
+          radiicopy.pop(i)
+          nextPh4 = Pharmacophore.Pharmacophore(featcopy, radiicopy)
+          nextPh4.applyRadiiToBounds()
+        else:
+          nextPh4 = Pharmacophore.Pharmacophore(featcopy)
+        self._recursive_map(matchcopy, mat, nextPh4)
+
+  def get_mappings(self, featFactory):
```

Comment:
I'm willing to accept the PR without this change being made, but it would be nice if the naming here were consistent with (most of) the rest of the RDKit. Please use CamelCase for method names.